### PR TITLE
Jupyter notebook test

### PR DESCRIPTION
**Pull request by PhilipMay**
First version of jupyter notebook test as discussed in #485

### CODE REVIEW

**Comment by PhilipMay** [comment]
There are still the following issues:
- the path (NBDIR) is hard wired and maybe wrong
- bazel integration
- is this the right place for this file?

**Comment by Smokrow** [comment]
I think the hardwired NBDIR should not be a problem since the folder should not move around after they included it into the docs generator. (see this [comment](https://github.com/tensorflow/addons/issues/485#issuecomment-529647505) ) 

I am not quite sure if we want to put this into the bazel testing pipeline since it is kind of self contained. We could add it to the travis pipeline though. @seanpmorgan Maybe we can add another test after the bazel build pipeline and before the build. 

@PhilipMay do you know if this testing environment runs `!pip install lalala` arguments?

File: tools/ci_testing/examples_test.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (seanpmorgan) [inline tools/ci_testing/examples_test.py]: Please make this `tutorials_test.py`

File: tools/ci_testing/examples_test.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Test for example jupyter notebooks."""
+
+import testipynb
```
Comment (seanpmorgan) [inline tools/ci_testing/examples_test.py]: We'll need a script to install this dependency during test time. Could you add this to:
https://github.com/tensorflow/addons/blob/master/tools/ci_build/install/ci_requirements.txt

File: tools/ci_testing/examples_test.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import unittest
+
+NBDIR = 'examples'
```
Comment (seanpmorgan) [inline tools/ci_testing/examples_test.py]: `docs/tutorials`

File: tools/ci_testing/examples_test.py (hunk 1 above)
Comment (PhilipMay) [reply tools/ci_testing/examples_test.py]: renamed it

File: tools/ci_testing/examples_test.py (hunk 2 above)
Comment (PhilipMay) [reply tools/ci_testing/examples_test.py]: Done - but without specifying a specific version number.
Is that ok?

**Comment by PhilipMay** [comment]
> @PhilipMay do you know if this testing environment runs `!pip install lalala` arguments?

Yes it runs pip this way. I uninstalled tensorflow with pip, started `make tutorial_test` and then checked if tensorflow got installed. It was...

**Comment by PhilipMay** [comment]
From my point of view this is done. These steps still miss:
- integrate to travis - do this in other PR?
- the test now shows errors so fix the notebooks - also separate PR?

File: makefile [hunk 4]
```diff
@@ -30,3 +30,6 @@ unit-test:
 
 gpu-unit-test:
 	bash tools/ci_testing/addons_gpu.sh
+
+tutorials-test:
```
Comment (seanpmorgan) [inline makefile]: can you make it so this install the dependencies.
```
tutorials-test: install-ci-dependency
    docker run --rm -it -v ${PWD}:/addons -w /addons tensorflow/tensorflow:custom-op-ubuntu16 /bin/bash
```

File: tools/ci_build/install/ci_requirements.txt [hunk 5]
```diff
@@ -2,3 +2,4 @@ pylint == 1.9 ; python_version <= '2.7'
 pylint == 2.3.1 ; python_version > '2.7'
 docformatter == 1.1
 yapf == 0.26.0
+testipynb
```
Comment (seanpmorgan) [inline tools/ci_build/install/ci_requirements.txt]: Could you pin it to `0.0.2`. Just makes maintenance easier.

File: makefile (hunk 4 above)
Comment (PhilipMay) [reply makefile]: I am confused a bit. How does it start the tests with your commands above?

**Comment by PhilipMay** [comment]
> Thanks @PhilipMay! I'm seeing several notebooks fail for `DeadKernelError: Kernel died` possibly from timeout. Could you increase the timeout for the tests and see if it fixes it.

@seanpmorgan I can see no `DeadKernelError: Kernel died` messages in my log (see below). The timeout is already increased from 600 (default) to 2100. Can you maybe post your messages? Where do these differences come from?

```
FAIL: test_optimizers_lazyadam (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... optimizers_lazyadam FAILED 

SyntaxError in cell [2] 
-----------
!pip install tensorflow-gpu==2.0.0rc0
!pip install tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function, unicode_literals

import tensorflow as tf
import tensorflow_addons as tfa
import tensorflow_datasets as tfds
import numpy as np
from matplotlib import pyplot as plt
-----------

                            

======================================================================
FAIL: test_template (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... template FAILED 

NameError in cell [5] 
-----------
result = model(tf.constant(np.random.randn(10,5), dtype = tf.float32)).numpy()

print("min:", result.min())
print("max:", result.max())
print("mean:", result.mean())
print("shape:", result.shape)
-----------

                            

----------------------------------------------------------------------
Ran 6 tests in 94.164s

FAILED (failures=6)
make: *** [tutorials-test] Error 1
(tf2) mikes-MacBook:addons mike$ clear

(tf2) mikes-MacBook:addons mike$ make tutorials-test
python tools/ci_testing/tutorials_test.py

--------------------- Testing image_ops.ipynb ---------------------
/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/jupyter_client/session.py:371: DeprecationWarning: Session._key_changed is deprecated in traitlets 4.1: use @observe and @unobserve instead.
  def _key_changed(self):


 ... image_ops FAILED 

SyntaxError in cell [2] 
-----------
!pip install -q tensorflow-gpu==2.0.0rc0
!pip install -q tensorflow-addons~=0.5

from __future__ import absolute_import, division, print_function, unicode_literals

import numpy as np
import tensorflow as tf
import tensorflow_addons as tfa

import matplotlib.pyplot as plt
-----------

                            
----------------- >> begin Traceback << ----------------- 

  File "<ipython-input-2-b46ec5485964>", line 7
    import tensorflow as tf
           ^
SyntaxError: from __future__ imports must occur at the beginning of the file



----------------- >> end Traceback << -----------------

                            
F
--------------------- Testing layers_normalizations.ipynb ---------------------


 ... layers_normalizations FAILED 

SyntaxError in cell [2] 
-----------
!pip install -q tensorflow==2.0.0rc0 
!pip install -q tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function
import tensorflow as tf
import tensorflow_addons as tfa
-----------

                            
----------------- >> begin Traceback << ----------------- 

  File "<ipython-input-2-ea1d9ed991dc>", line 6
SyntaxError: from __future__ imports must occur at the beginning of the file



----------------- >> end Traceback << -----------------

                            
F
--------------------- Testing layers_weightnormalization.ipynb ---------------------


 ... layers_weightnormalization FAILED 

SyntaxError in cell [2] 
-----------
!pip install tensorflow-gpu==2.0.0rc0
!pip install tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function, unicode_literals

import tensorflow as tf
import tensorflow_addons as tfa
import numpy as np
from matplotlib import pyplot as plt
-----------

                            
----------------- >> begin Traceback << ----------------- 

  File "<ipython-input-2-6458efa20296>", line 6
    import tensorflow_addons as tfa
           ^
SyntaxError: from __future__ imports must occur at the beginning of the file



----------------- >> end Traceback << -----------------

                            
F
--------------------- Testing losses_triplet.ipynb ---------------------


 ... losses_triplet FAILED 

SyntaxError in cell [2] 
-----------
!pip install -q tensorflow-gpu==2.0.0rc0
!pip install -q tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function, unicode_literals

import io
import numpy as np
import tensorflow as tf
import tensorflow_addons as tfa
import tensorflow_datasets as tfds
from matplotlib import pyplot as plt
from google.colab import files
-----------

                            
----------------- >> begin Traceback << ----------------- 

  File "<ipython-input-2-aeb553bdcf4f>", line 6
    import numpy as np
           ^
SyntaxError: from __future__ imports must occur at the beginning of the file



----------------- >> end Traceback << -----------------

                            
F
--------------------- Testing optimizers_lazyadam.ipynb ---------------------


 ... optimizers_lazyadam FAILED 

SyntaxError in cell [2] 
-----------
!pip install tensorflow-gpu==2.0.0rc0
!pip install tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function, unicode_literals

import tensorflow as tf
import tensorflow_addons as tfa
import tensorflow_datasets as tfds
import numpy as np
from matplotlib import pyplot as plt
-----------

                            
----------------- >> begin Traceback << ----------------- 

  File "<ipython-input-2-444c83055f39>", line 6
    import tensorflow_addons as tfa
           ^
SyntaxError: from __future__ imports must occur at the beginning of the file



----------------- >> end Traceback << -----------------

                            
F
--------------------- Testing template.ipynb ---------------------
2019-09-25 16:20:27.875261: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-09-25 16:20:27.899003: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7fd57d5aa130 executing computations on platform Host. Devices:
2019-09-25 16:20:27.899078: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/logging/__init__.py", line 1945, in shutdown
    h.flush()
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/absl/logging/__init__.py", line 891, in flush
    self._current_handler.flush()
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/absl/logging/__init__.py", line 785, in flush
    self.stream.flush()
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/ipykernel/iostream.py", line 341, in flush
    if self.pub_thread.thread.is_alive():
AttributeError: 'NoneType' object has no attribute 'thread'


 ... template FAILED 

NameError in cell [5] 
-----------
result = model(tf.constant(np.random.randn(10,5), dtype = tf.float32)).numpy()

print("min:", result.min())
print("max:", result.max())
print("mean:", result.mean())
print("shape:", result.shape)
-----------

                            
----------------- >> begin Traceback << ----------------- 

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-5-3ed927615e9f> in <module>
----> 1 result = model(tf.constant(np.random.randn(10,5), dtype = tf.float32)).numpy()
      2 
      3 print("min:", result.min())
      4 print("max:", result.max())
      5 print("mean:", result.mean())

NameError: name 'np' is not defined


----------------- >> end Traceback << -----------------

                            
F
======================================================================
FAIL: test_image_ops (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... image_ops FAILED 

SyntaxError in cell [2] 
-----------
!pip install -q tensorflow-gpu==2.0.0rc0
!pip install -q tensorflow-addons~=0.5

from __future__ import absolute_import, division, print_function, unicode_literals

import numpy as np
import tensorflow as tf
import tensorflow_addons as tfa

import matplotlib.pyplot as plt
-----------

                            

======================================================================
FAIL: test_layers_normalizations (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... layers_normalizations FAILED 

SyntaxError in cell [2] 
-----------
!pip install -q tensorflow==2.0.0rc0 
!pip install -q tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function
import tensorflow as tf
import tensorflow_addons as tfa
-----------

                            

======================================================================
FAIL: test_layers_weightnormalization (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... layers_weightnormalization FAILED 

SyntaxError in cell [2] 
-----------
!pip install tensorflow-gpu==2.0.0rc0
!pip install tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function, unicode_literals

import tensorflow as tf
import tensorflow_addons as tfa
import numpy as np
from matplotlib import pyplot as plt
-----------

                            

======================================================================
FAIL: test_losses_triplet (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... losses_triplet FAILED 

SyntaxError in cell [2] 
-----------
!pip install -q tensorflow-gpu==2.0.0rc0
!pip install -q tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function, unicode_literals

import io
import numpy as np
import tensorflow as tf
import tensorflow_addons as tfa
import tensorflow_datasets as tfds
from matplotlib import pyplot as plt
from google.colab import files
-----------

                            

======================================================================
FAIL: test_optimizers_lazyadam (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... optimizers_lazyadam FAILED 

SyntaxError in cell [2] 
-----------
!pip install tensorflow-gpu==2.0.0rc0
!pip install tensorflow-addons~=0.5
from __future__ import absolute_import, division, print_function, unicode_literals

import tensorflow as tf
import tensorflow_addons as tfa
import tensorflow_datasets as tfds
import numpy as np
from matplotlib import pyplot as plt
-----------

                            

======================================================================
FAIL: test_template (testipynb.testipynb.NbTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/miniconda3/envs/tf2/lib/python3.6/site-packages/testipynb/testipynb.py", line 96, in test_func
    assert passing, msg
AssertionError: 

 ... template FAILED 

NameError in cell [5] 
-----------
result = model(tf.constant(np.random.randn(10,5), dtype = tf.float32)).numpy()

print("min:", result.min())
print("max:", result.max())
print("mean:", result.mean())
print("shape:", result.shape)
-----------

                            

----------------------------------------------------------------------
Ran 6 tests in 96.071s

FAILED (failures=6)
make: *** [tutorials-test] Error 1
```

File: tools/ci_build/install/ci_requirements.txt (hunk 5 above)
Comment (PhilipMay) [reply tools/ci_build/install/ci_requirements.txt]: done that

File: makefile (hunk 4 above)
Comment (seanpmorgan) [reply makefile]: Ah apologies I copied the wrong command in the bottom line. Please see the edited block

Comment (PhilipMay) [reply makefile]: Ok. Pushed the update.

**Comment by PhilipMay** [comment]
@seanpmorgan what do you think are the next steps? Fix the `DeadKernelError: Kernel died ` issue you have and I can't reproduce - right?

**Comment by seanpmorgan** [comment]
Hi Philip could you please verify how you're testing the code? With the current commit I'm getting:
```
root@a4cbd4d4bf35:/addons# make tutorials-test
makefile:35: *** missing separator.  Stop
```
The indentation must be consistent with the other commands. 

Also, could you please run in the docker container so we have consistent environments and that is how this will be ran during automated testing.

For a non-gpu environment I'm getting a mix of issues related to #532 (which is unrelated to this) and DeadKernel errors:
[tutorialtest_output.txt](https://github.com/tensorflow/addons/files/3657766/tutorialtest_output.txt)


When running in gpu environment just to mitigate the #532 issues I get 
[tutorialtest_gpu_output.txt](https://github.com/tensorflow/addons/files/3657779/tutorialtest_gpu_output.txt)

As you can see a number of imports fail (matplotlib, tensorflow_datasets, numpy). So what we'll need to do is create a separate `install_nbtest_dependecy.sh` script which runs for this make command.

We can't merge this until we know that the setup is capable of running these tests

**Comment by PhilipMay** [comment]
> Hi Philip could you please verify how you're testing the code? With the current commit I'm getting:
> 
> ```
> root@a4cbd4d4bf35:/addons# make tutorials-test
> makefile:35: *** missing separator.  Stop
> ```

This is fixed now.

**Comment by Smokrow** [comment]
> As you can see a number of imports fail (matplotlib, tensorflow_datasets, numpy). So what we'll need to do is create a separate `install_nbtest_dependecy.sh` script which runs for this make command.

@seanpmorgan Do you think we could use the Colab Notebook container as an environment since this is the environment people will actually see this in? I am not quite sure if they are public though. Maybe a googler can help out? @karmel

**Comment by karmel** [comment]
I don't quite follow-- what colab nb container are you referring to, exactly?

**Comment by Smokrow** [comment]
I always thought that the "standard" Google Colab Notebook is actually a container running on Googles Servers. In case it is, do you know if the container is public?

**Comment by karmel** [comment]
@MarkDaoust @yashk2810 @tomerk , who know more about how the nb tests are run and what the options here might be.

**Comment by yashk2810** [comment]
One option is, since we are creating a subsite for addons, their notebooks will automatically get tested by the infrastructure we have in place for all the other subsites and TensorFlow. If the notebooks fail, an email will be sent to the addons team to look into what went wrong.

For CI testing, I can look into opensourcing the stuff that we use to test the notebooks.

@seanpmorgan @PhilipMay One thing to note for enabling the tests with our infra is to not save the outputs in the notebook and install whichever version of addons you want to test your notebook with. You can follow the template notebook (https://github.com/tensorflow/docs/blob/master/tools/templates/notebook.ipynb).

**Comment by facaiy** [comment]
> For CI testing, I can look into opensourcing the stuff that we use to test the notebooks.

Good news! Thank Yash. It sound like we can share the same notebook testing library (and even workflow) if opensourcing. Would you mind telling us the opensouring schedule if possible?

**Comment by yashk2810** [comment]
> Would you mind telling us the opensouring schedule if possible?

Probably within a few weeks :)

**Comment by PhilipMay** [comment]
Hi @yashk2810 - what is the status here? 
Thanks
Philip

**Comment by PhilipMay** [comment]
Bumping it again... @yashk2810 @karmel

**Comment by yashk2810** [comment]
Hi @PhilipMay,

Opensourcing what we use will take some time since it needs to be decoupled from a lot of things that we use and will require some effort. I will triage and set a priority on it. But you are free to explore any other alternative that you have in mind :)

On a positive note, since addons is now a subsite tensorflow.org/addons, every notebook is already getting tested. If something fails, we will send an email to the appropriate handle.

**Comment by lc0** [comment]
@seanpmorgan does it mean, that the issue is redundant and could be closed?

**Comment by seanpmorgan** [comment]
IMO yes, provided that the above system works. @yashk2810 What email are failures sent to?
@PhilipMay would this be an acceptable solution?

**Comment by PhilipMay** [comment]
> @PhilipMay would this be an acceptable solution?

For me this is 100% ok. Why not. No need to implement something that is already implemented.

> On a positive note, since addons is now a subsite tensorflow.org/addons, every notebook is already getting tested. If something fails, we will send an email to the appropriate handle.

Are we sure that "send an email to the appropriate handle" does work and that this will always be converted to a ticket here in GitHub?

**Comment by yashk2810** [comment]
> always be converted to a ticket here in GitHub?
I don;t think it will be converted to a ticket on Github.

But, is there an email with all the maintainers of addons that we can forward it to? Then you can monitor that email for failure emails from our side.

**Comment by PhilipMay** [comment]
> But, is there an email with all the maintainers of addons that we can forward it to? Then you can monitor that email for failure emails from our side.

I dont know... @seanpmorgan do you?
Before we close this PR we should test the mail functionality.

**Comment by lc0** [comment]
Also, can it be on CI level? So we get the alert before merging to master
and running nightly builds.

**Comment by karmel** [comment]
(@yashk2810 -- not sure how these particular tests are set up, but when I previously tried in several ways to get internal testing emails forwarded to external aliases, I failed. We could possibly set up an internal alias with external members or something like that, but note that there may be some hoops to jump through.)

**Comment by yashk2810** [comment]
> We could possibly set up an internal alias

Yup, that's how it works. The email is sent to our internal alias and then we forward it to the appropriate external handle after scrubbing all the internal email id's.

We have been doing this for translations already.

**Comment by seanpmorgan** [comment]
@yashk2810 The email for failing tests is: `addons-testing@tensorflow.org`. It's open to all as a [google group](https://groups.google.com/a/tensorflow.org/forum/#!forum/addons-testing).

 Do you have any more information as to when the notebooks are tested? Is this only done once the API docs are re-generated after every release? We'll be releasing 0.7 in the next few days. I can put a broken tutorial as a test on the branch.

**Comment by yashk2810** [comment]
> The email for failing tests is: addons-testing@tensorflow.org

Thanks

> done once the API docs are re-generated after every release

Yes, currently. If the changes made to the notebooks are very often, then I can move it to weekly and eventually nightly. 

> I can put a broken tutorial as a test on the branch

If you want to test it out, sure. But we pull docs from the master branch. So you won;t have to put it on the 0.7 branch :)

**Comment by seanpmorgan** [comment]
> > The email for failing tests is: [addons-testing@tensorflow.org](mailto:addons-testing@tensorflow.org)
> 
> Thanks
> 
> > done once the API docs are re-generated after every release
> 
> Yes, currently. If the changes made to the notebooks are very often, then I can move it to weekly and eventually nightly.
> 
> > I can put a broken tutorial as a test on the branch
> 
> If you want to test it out, sure. But we pull docs from the master branch. So you won;t have to put it on the 0.7 branch :)

Great thanks for your help @yashk2810! We'll put up a test prior to requesting the next update of API docs. 

I did want to confirm that addons docs are pulled from master though. We only generate API docs on our release branches so can you confirm that the addons site generates them from master branch separately from what we do?

**Comment by yashk2810** [comment]
> addons site generates them from master branch

So, for the narrative docs (guides, tutorials, etc) we pull from master.

For API docs we use your pip package to generate the api docs.

Does that make sense?

**Comment by seanpmorgan** [comment]
Thanks all for the support on this! Closing this PR and created a subsequent issue to track this testing:
https://github.com/tensorflow/addons/issues/768

**Pull request closed without merging**